Add a tutorial (#25)
Qiskit-Extensions/arraylias: docs/tutorials/schrödinger_equation.rst

.. _rungekutta:

.. module:: rungekutta

Solving the Schrodinger equation with either NumPy or JAX using the default NumPy alias
=======================================================================================

In this tutorial, we will learn how to write generic code with examples of numerical computation by using 
`NumPy <https://numpy.org/>`_ and `JAX <https://github.com/google/jax>`_.
We'll use a code example that employs the Runge-Kutta method to numerically simulate the matrix evolution of the Schrödinger equation.

> In addition, we will showcase the utility of :meth:`.Alias.register_function` method for efficient problem-solving.

Reviewer: ```suggestion
In addition, we will show how to add new dispatched function to :class:`.Alias` instance using :meth:`.Alias.register_function`.
```

In this section, we will go through the following steps:

> 1. Import the required libraries and initialize the numpy alias.

Reviewer: ```suggestion
1. Import the required libraries and initialize the default NumPy alias.
```

> 2. Define the equation using :class:`.Alias`.

Reviewer: ```suggestion
2. Define a function for evaluating the right-hand side of the Schrodinger equation using using the default NumPy alias.
```

> 3. Solve the right-hand side function using existing solvers.

Reviewer: ```suggestion
3. Solve the Schrodinger equation using the right-hand side function and existing NumPy-based and JAX-based solvers.
```

> 4. Register custom solvers using :meth:`.Alias.register_function` method.

Reviewer: ```suggestion
4. Register new dispatched custom solvers using :meth:`.Alias.register_function` method.
```

> 5. Solve using the custom function.

Reviewer: ```suggestion
5. Solve the Schrodinger equation using the custom solvers.
```

1. Import the required libraries and initialize the numpy alias
---------------------------------------------------------------

> Here, we import the necesary libraries.

Reviewer: ```suggestion
Here, we import the necessary libraries.
```

    import numpy as np
    import matplotlib.pyplot as plt
    import jax
    import jax.numpy as jnp

    from scipy.integrate import solve_ivp
    from jax.experimental.ode import odeint

    # set some variables
    dt = 0.01
    N = 1001

Initialize the default NumPy alias.

.. jupyter-execute::

    alias = numpy_alias()
    unp = alias()

2. Define the equation using :class:`.Alias`
--------------------------------------------

We solve the Schrödinger equation using the Runge-Kutta method in this tutorial.
The Schrödinger equation is the differential equation

.. math:: \psi'(t) = -i * H(t) \psi(t),

where :math:`H(t)` is a time-dependent matrix called the Hamiltonian, and :math:`\psi(t)` is the state of the system.

We will solve a common model for a two-level quantum system, which has Hamiltonian

.. math:: H(t) = \times 2 \pi \nu_z \frac{Z}{2} + 2 \pi \nu_x \cos(2 \pi \nu_d t)\frac{X}{2},

where :math:`\{X,Y,Z\}` are the Pauli matrices, and :math:`\nu_z`, :math:`\nu_x`, and :math:`\nu_d` are model parameters of the system.

Write a function representing the right-hand side of the Schrodinger equation with the above Hamiltonian.

.. jupyter-execute::

>     Z = np.array([[1,0],[0,-1]])
>     X = np.array([[0,1],[1,0]])

Reviewer: ```suggestion
    Z = np.array([[1, 0], [0, -1]])
    X = np.array([[0, 1], [1, 0]])
```

>     def rhs(t,y):

Reviewer: ```suggestion
    def rhs(t, y):
```

        return unp.matmul(-1j * (5 * Z -  unp.cos(10 * t) * X ), y)

> We can confirm that the rhs function outputs the type corresponding to the input type.

Reviewer: ```suggestion
Depending on the input type, the `rhs` function will execute using either NumPy or JAX, which we can confirm by observing the output types:
```

.. jupyter-execute::

    # Numpy input
    rhs(0.1, np.array([0., 1.]))

.. jupyter-execute::

    # Jax.numpy input
    rhs(jnp.array(0.1), jnp.array([0., 1.]))

> We eventually want to find the probability of existence of this qubit state, so we will prepare the following function.

Reviewer: ```suggestion
Define a function for computing the probability of observing the system in a given state to be used throughout the tutorial:
```

.. jupyter-execute::

    def state_probabilities(state):
        return unp.abs(state) ** 2

3. Solve the right-hand side function using existing solvers
------------------------------------------------------------

Here we show how the rhs function can be passed to numerical ODE solvers in both SciPy and JAX as if the function had been natively written in either library.

First, we solve the equation by using Numpy as the input and ``scipy.integrate.solve_ivp`` as a solver.
We define the initial state, the time span for the simulation, and time point we want to simulate.

.. jupyter-execute::

    init_state = np.array([1. + 0j,0. + 0j])

    t_span = [0,(N-1) * dt]

>     T = np.linspace(0,(N-1) * dt,N)

Reviewer: ```suggestion
    T = np.linspace(0, (N-1) * dt,N)
```

We solve by using ``scipy.integrate.solve_ivp`` and plot the probabilities of each state.

.. jupyter-execute::

>     sol = solve_ivp(rhs,t_span,init_state,method='RK45',t_eval=T)

Reviewer: ```suggestion
    sol = solve_ivp(rhs, t_span, init_state, method='RK45', t_eval=T)
```

    probabilities = state_probabilities(sol.y)

    plt.plot(sol.t, probabilities[0], label="0")
    plt.plot(sol.t, probabilities[1], label="1")
    plt.xlabel('T')
    plt.ylabel('Probability')
    plt.legend()
    plt.show()

Second, we solve the equation by using Jax.array as the input and ``jax.experimental.ode.odeint`` as a solver.

.. jupyter-execute::

    init_state = jnp.array([1. + 0j,0. + 0j])

>     t_span = [0,(N-1) * dt]
>     T = jnp.linspace(0,(N-1) * dt,N)

Reviewer: ```suggestion
    t_span = [0, (N-1) * dt]
    T = jnp.linspace(0, (N-1) * dt, N)
```

>     sol = odeint(lambda y,t: rhs(t,y), init_state, T)

Reviewer: ```suggestion
    sol = odeint(lambda y, t: rhs(t, y), init_state, T)
```

    probabilities = state_probabilities(sol.T)
    plt.plot(T, probabilities[0], label="0")
    plt.plot(T, probabilities[1], label="1")
    plt.xlabel('T')
    plt.ylabel('Probability')
    plt.legend()
    plt.show()

4. Register custom solvers using :meth:`.Alias.register_function`
-----------------------------------------------------------------

In this section, we define custom functions for solving differential equations in both NumPy and JAX, and register them to our instance of the NumPy alias.

We will use the 4th order Runge-Kutta method, whose single step formula is:

.. math::

    k_1 &= h \cdot f(t_n, y_n) \\
    k_2 &= h \cdot f(t_n + \frac{h}{2}, y_n + \frac{k_1}{2}) \\
    k_3 &= h \cdot f(t_n + \frac{h}{2}, y_n + \frac{k_2}{2}) \\
    k_4 &= h \cdot f(t_n + h, y_n + k_3) \\
    y_{n+1} &= y_n + \frac{1}{6}(k_1 + 2k_2 + 2k_3 + k_4)

> , where :math:`y_{n}`, :math:`t_{n}`, and :math:`h` are current solution, current time, and time step size, respectively.

Reviewer: ```suggestion
where :math:`y_{n}`, :math:`t_{n}`, and :math:`h` are current solution, current time, and time step size, respectively.
```

> We define the Runge-Kutta method to be used later here:

Reviewer: ```suggestion
Define a function for taking a single Runge-Kutta step:
```

.. jupyter-execute::

    def runge_kutta_step(t, y, dt, rhs):
        k1 = dt * rhs(t, y)
        k2 = dt * rhs(t + 0.5 * dt, y + 0.5*k1)
        k3 = dt * rhs(t + 0.5 * dt, y + 0.5*k2)
        k4 = dt * rhs(t + dt, y + k3)

>         return (k1 + 2*k2 + 2*k3 + k4) / 6.

Reviewer: ```suggestion
        return (k1 + 2 * k2 + 2 * k3 + k4) / 6.
```

First, define the version of the solver written for use with standard NumPy, and register it to our ``alias`` instance to act on NumPy arrays using ``alias.register_function``:

.. jupyter-execute::

    @alias.register_function(lib="numpy", path="runge_kutta")
    def _(y0, dt, N, rhs):
        probabilities = []
        for n in range(N):
            probabilities.append(state_probabilities(y0))

>             y0+= runge_kutta_step(n * dt, y0, dt, rhs)

Reviewer: ```suggestion
            y0 += runge_kutta_step(n * dt, y0, dt, rhs)
```

        return probabilities

Next, register a version of the solver to work on JAX arrays. For better behaviour under JAX transformations, we need to use the JAX looping construct ``jax.lax.scan`` rather than the standard Python ``for`` loop:

.. jupyter-execute::

    @alias.register_function(lib="jax", path="runge_kutta")
    def _(y0, dt, N, rhs):
        def runge_kutta_step_scan(carry, probabilities):
            n, y = carry
            probabilities = state_probabilities(y)
            y+= runge_kutta_step(n * dt, y, dt, rhs)
            return (n + 1, y), probabilities
        _, probabilities = jax.lax.scan(runge_kutta_step_scan, (0, y0), jnp.zeros((N,2)))
        return probabilities

5. Solve using the custom function
----------------------------------

Finally, we will solve the Schrödinger equation using both the NumPy and JAX libraries via our single dispatched function ``unp.runge_kutta``.

First, solve with NumPy:

.. jupyter-execute::

    init_state = np.array([1. + 0j,0. + 0j])

    probabilities = unp.array(unp.runge_kutta(init_state, dt, N, rhs))

>     T = np.linspace(0,(N-1) * dt,N)

Reviewer: ```suggestion
    T = np.linspace(0, (N-1) * dt, N)
```

    plt.plot(T, probabilities, label = ["0", "1"])
    plt.xlabel('T')
    plt.ylabel('Probability')
    plt.legend()
    plt.show()

.. jupyter-execute::

    %timeit unp.array(unp.runge_kutta(init_state, dt, N, rhs))

Second case is JAX:

.. jupyter-execute::

>     init_state = jnp.array([1. + 0j,0. + 0j])

Reviewer: ```suggestion
    init_state = jnp.array([1. + 0j, 0. + 0j])
```

    probabilities = unp.array(unp.runge_kutta(init_state, dt, N, rhs))

>     T = np.linspace(0,(N-1) * dt,N)

Reviewer: ```suggestion
    T = np.linspace(0, (N-1) * dt, N)
```

    plt.plot(T, probabilities, label=["0", "1"])
    plt.xlabel('T')
    plt.ylabel('Probability')
    plt.legend()
    plt.show()

Lastly, we verify that the function ``unp.runge_kutta`` behaves as expected under JAX transformations.

.. jupyter-execute::

    from functools import partial

>     @partial(jax.jit, static_argnums=(2,3))

Reviewer: ```suggestion
    @partial(jax.jit, static_argnums=(2, 3))
```

.. jupyter-execute::

    init_state = np.array([1. + 0j,0. + 0j])

    t_span = [0,(N-1) * dt]
    T = np.linspace(0,(N-1) * dt,N)

We solve by using ``scipy.integrate.solve_ivp`` and plot the probabilities of each state.

.. jupyter-execute::

    sol = solve_ivp(rhs,t_span,init_state,method='RK45',t_eval=T)
    probabilities = state_probabilities(sol.y)

    plt.plot(sol.t, probabilities[0], label="0")
    plt.plot(sol.t, probabilities[1], label="1")
    plt.xlabel('T')
    plt.ylabel('Probability')
    plt.legend()
    plt.show()

Second, we solve the equation by using Jax.array as the input and ``jax.experimental.ode.odeint`` as a solver.

.. jupyter-execute::

    init_state = jnp.array([1. + 0j,0. + 0j])

    t_span = [0,(N-1) * dt]
    T = jnp.linspace(0,(N-1) * dt,N)

    sol = odeint(lambda y,t: rhs(t,y), init_state, T)
    probabilities = state_probabilities(sol.T)
    plt.plot(T, probabilities[0], label="0")
    plt.plot(T, probabilities[1], label="1")
    plt.xlabel('T')
    plt.ylabel('Probability')
    plt.legend()
    plt.show()

4. Register custom solvers using :meth:`.Alias.register_function`
-----------------------------------------------------------------

> In this section, we introduce how to register the function, taking the 4th order Runge-Kutta method
> to solve ordinary differential equations as an example.

Reviewer: ```suggestion
In this section, we define custom functions for solving differential equations in both NumPy and JAX, and register them to our instance of the NumPy alias.
```

> The 4th order Runge-Kutta fomula is here:

Reviewer: ```suggestion
We will use the 4th order Runge-Kutta method, whose single step formula is:
```

.. math::

    k_1 &= h \cdot f(t_n, y_n) \\
    k_2 &= h \cdot f(t_n + \frac{h}{2}, y_n + \frac{k_1}{2}) \\
    k_3 &= h \cdot f(t_n + \frac{h}{2}, y_n + \frac{k_2}{2}) \\
    k_4 &= h \cdot f(t_n + h, y_n + k_3) \\
    y_{n+1} &= y_n + \frac{1}{6}(k_1 + 2k_2 + 2k_3 + k_4)

, where :math:`y_{n}`, :math:`t_{n}`, and :math:`h` are current solution, current time, and time step size, respectively.

We define the Runge-Kutta method to be used later here:

.. jupyter-execute::

    def runge_kutta_step(n, state):
        k1 = dt * rhs(n * dt, state)
        k2 = dt * rhs(n * dt + 0.5 * dt, state + 0.5*k1)
        k3 = dt * rhs(n * dt + 0.5 * dt, state + 0.5*k2)
        k4 = dt * rhs(n * dt + dt, state + k3)

>         return (k1 + 2*k2 + 2*k3 + k4) / 6.

Reviewer: I wonder if here the method should be defined to take in the state, current time, time step size, and rhs. E.g.

```
def runge_kutta_step(t, y, dt, rhs):
    k1 = dt * rhs(t, y)
    k2 = dt * rhs(t + 0.5 * dt, y + 0.5 * k1)
    ...
```

> When writing the custom solver, we write and register a version for both NumPy and JAX. 
> The NumPy version uses standard Python loops, whereas the JAX version uses the JAX looping construct ``jax.lax.scan``.

Reviewer: I think this explanation can be removed, and `jax.lax.scan` can be referenced before you define the JAX version.

```suggestion
```

> In Numpy case, we define the function ``runge_kutta`` for Numpy.

Reviewer: ```suggestion
First, define the version of the solver written for use with standard NumPy, and register it to our ``alias`` instance to act on NumPy arrays using ``alias.register_function``:
```

.. jupyter-execute::

    @alias.register_function(lib="numpy", path="runge_kutta")
    def _(state, N):
        probabilities = []
        for n in range(N):
            probabilities.append(state_probabilities(state))
            state+= runge_kutta_step(n, state)

>         return probabilities

Reviewer: In line with the above comment about defining `runge_kutta_step` to have signature `runge_kutta_step(y, t, dt, rhs)`, I think this could be rewritten to have a signature like

```
def _(y0, t0, dt, N, rhs):
   ....
```
It's not a huge deal but this is the more common way in which the interface for DE solvers are written - an argument for the initial state, some way of specifying the solution interval, and the passed RHS function.

Author: I changed at https://github.com/Qiskit-Extensions/arraylias/pull/25/commits/a5e84316f9712bcb652813d1494a97185cfd8d76.

> This custom function of ``runge_kutta`` is registered using the decorator ``@alias.register_function`` under 
> NumPy.
> 

Reviewer: ```suggestion
```

> In the case of JAX,

Reviewer: ```suggestion
Next, register a version of the solver to work on JAX arrays. For better behaviour under JAX transformations, we need to use the JAX looping construct ``jax.lax.scan`` rather than the standard Python ``for`` loop:
```

.. jupyter-execute::

    @alias.register_function(lib="jax", path="runge_kutta")
    def _(state, N):
        def runge_kutta_step_scan(carry, probabilities):
            n, state = carry
            probabilities = state_probabilities(state)
            state+= runge_kutta_step(n, state)
            return (n + 1, state), probabilities
        _, probabilities = jax.lax.scan(runge_kutta_step_scan, (0, state), jnp.zeros((N,2)))
        return probabilities

5. Solve using the custom function
----------------------------------

> We have just completed writing the generic code to solve the Schrödinger equation.
> We will now demonstrate two cases of solving the equation using NumPy and JAX as inputs.

Reviewer: ```suggestion
Finally, we will solve the Schrödinger equation using both the NumPy and JAX libraries via our single dispatched function ``unp.runge_kutta``.
```

> The Numpy case is here:

Reviewer: ```suggestion
First, solve with NumPy:
```

.. jupyter-execute::

    init_state = np.array([1. + 0j,0. + 0j])

    probabilities = unp.array(unp.runge_kutta(init_state, N))

    T = np.linspace(0,(N-1) * dt,N)
    plt.plot(T, probabilities, label = ["0", "1"])
    plt.xlabel('T')
    plt.ylabel('Probability')
    plt.legend()
    plt.show()

.. jupyter-execute::

    %timeit unp.array(unp.runge_kutta(init_state, N))

Second case is JAX:

.. jupyter-execute::

    init_state = jnp.array([1. + 0j,0. + 0j])
    probabilities = unp.array(unp.runge_kutta(init_state, N))

    T = np.linspace(0,(N-1) * dt,N)

    plt.plot(T, probabilities, label=["0", "1"])
    plt.xlabel('T')
    plt.ylabel('Probability')
    plt.legend()
    plt.show()

> We see if we can actually jit the custom solver.

Reviewer: ```suggestion
Lastly, we verify that the function ``unp.runge_kutta`` behaves as expected under JAX transformations.
```

.. _rungekutta:

.. module:: rungekutta

Time evolution of the Schrödinger equation

> ==========================================

Author: I changed the title at https://github.com/Qiskit-Extensions/arraylias/pull/25/commits/145a5bfe95d0fd6bd4c883b56673455a05f6a1b9.

Arraylias provides a way to configure aliased modules for certain libraries, thereby eliminating the need for us

> to write separate code for each of them.

Author: It is corresponding to the section that "Provide the minimum necessary explanation" in https://documentation.divio.com/.
I fixed at  https://github.com/Qiskit-Extensions/arraylias/pull/25/commits/2bb3a8e11aeac85bfa169ac1e310346ded62cc93 .

In this tutorial, we will learn how to write generic code with examples of numerical computation by using 
`NumPy <https://numpy.org/>`_ and `JAX <https://github.com/google/jax>`_.
We'll use a code example that employs the Runge-Kutta method to numerically simulate the matrix evolution of the Schrödinger equation.
In addition, we will showcase the utility of :meth:`.Alias.register_function` method for efficient problem-solving.

In this section, we will go through the following steps:

1. Import the required libraries.

> 2. Initialize the numpy alias.

Author: You're right. I fixed at https://github.com/Qiskit-Extensions/arraylias/pull/25/commits/724b55a9e30dadab827792084773cef65b270c68 .

> 3. Define the equation using :class:`.Alias`.

Author: I fixed at https://github.com/Qiskit-Extensions/arraylias/pull/25/commits/53eee4341415b76857707888dd7a1d1cf73d19c9 .

setting a bridge between these libraries.

We initialize Arraylias using the :func:`.numpy_alias` function and :class:`.Alias`.

.. jupyter-execute::

    alias = numpy_alias()
    unp = alias()

3. Define the equation using :class:`.Alias`
--------------------------------------------

We solve the Schrödinger equation using the Runge-Kutta method in this tutorial.
The Schrödinger equation is written as

.. math:: \frac{\partial\psi}{\partial t} = -i * H \psi

The equation we are solving represents a qubit's state as a two-level system. The Hamiltonian is

.. math:: H = \frac{1}{2} \times 2 \pi \nu_z {Z} + 2 \pi \nu_x \cos(2 \pi \nu_d t){X},

where :math:`\{X,Y,Z\}` are the Pauli matrices.

We can express the right-hand side (RHS) function of this equation as follows:

.. jupyter-execute::

    def rhs(t,y,Z,X):
        return unp.matmul(-1j * (5 * Z -  unp.cos(10 * t) * X ),y)

We can utilize :class:`.Alias` intuitively just like `NumPy <https://numpy.org/>`_, `JAX <https://github.com/google/jax>`_, and so on.
This generic code provided by Arraylias includes ``unp.matmul()``, which automatically select the appropriate module
based on the input type.

> 

Author: Certainly, it is important to provide the minimum explanation and show the concrete steps.
I added at https://github.com/Qiskit-Extensions/arraylias/pull/25/commits/e15965b67eb4e02788506b73629cf754a44e947f.

We eventually want to find the probability of existence of this qubit state, so we will prepare the following function.

.. jupyter-execute::

    def state_probabilities(state):
        probabilities = unp.abs(state) ** 2
        return probabilities / unp.sum(probabilities)

4. Solve the equation using existing solvers
--------------------------------------------

> 

Author: I added at https://github.com/Qiskit-Extensions/arraylias/pull/25/commits/72bd010114df00d1eb8c7184494a4a29494d807a.

    Z = np.array([[1,0],[0,-1]])
    X = np.array([[0,1],[1,0]])
    init_state = np.array([1. + 0j,0. + 0j])

    t_span = [0,(N-1) * dt]
    T = np.linspace(0,(N-1) * dt,N)

We solve by using ``scipy.integrate.solve_ivp`` and plot the probabilities of each state.

.. jupyter-execute::

    sol = solve_ivp(rhs,t_span,init_state,args=(Z,X),method='RK45',t_eval=T)
    probabilities = state_probabilities(sol.y)

    plt.plot(sol.t, probabilities[0], label="0")
    plt.plot(sol.t, probabilities[1], label="1")
    plt.xlabel('T')
    plt.ylabel('Probability')
    plt.legend()
    plt.show()

Second, we solve the equation by using Jax.array as the input and ``jax.experimental.ode.odeint`` as a solver.

.. jupyter-execute::

    Z = jnp.array([[1,0],[0,-1]])
    X = jnp.array([[0,1],[1,0]])
    init_state = jnp.array([1. + 0j,0. + 0j])

    t_span = [0,(N-1) * dt]
    T = jnp.linspace(0,(N-1) * dt,N)

    sol = odeint(lambda y,t: rhs(t,y,Z,X),init_state,T)
    probabilities = state_probabilities(sol.T)
    plt.plot(T, probabilities[0], label="0")
    plt.plot(T, probabilities[1], label="1")
    plt.xlabel('T')
    plt.ylabel('Probability')
    plt.legend()
    plt.show()

> 5. Register custom functions using :meth:`.Alias.register_function`

Reviewer: Maybe change this to 
```suggestion
5. Register custom solvers using :meth:`.Alias.register_function`
```

Author: I fixed at https://github.com/Qiskit-Extensions/arraylias/pull/25/commits/53eee4341415b76857707888dd7a1d1cf73d19c9 .

-------------------------------------------------------------------

Arraylais provides to register your own custom function using :meth:`.Alias.register_function` method.
In this section, we introduce how to register the function, taking the Runge-Kutta method as an example.

The Runge-Kutta method solves differential equations by approximating power series up to
the foruth order terms. It allows to find solutions through numerically calculations.

> We define the function for the Runge-Kutta method to be used later here:

Reviewer: I think here you can just write down the 4th order RK formula - no need to explain what `register_function` does, or explain what the Runge-Kutta method is for. You can just refer to it as a 4th order fixed step ODE method with formula: ....

Author: I updated at https://github.com/Qiskit-Extensions/arraylias/pull/25/commits/328ab5b771dca327efab87a6d5d6a871184d06de.

.. jupyter-execute::

    def runge_kutta_step(Z, X, n, state):
        k1 = dt * rhs(n * dt, state, Z, X)
        k2 = dt * rhs(n * dt + 0.5 * dt, state + 0.5*k1, Z, X)
        k3 = dt * rhs(n * dt + 0.5 * dt, state + 0.5*k2, Z, X)
        k4 = dt * rhs(n * dt + dt, state + k3, Z, X)
        return (k1 + 2*k2 + 2*k3 + k4) / 6.

JAX's strength lies in vectorized operations and parallel computations. In some cases, Python for loops
may not effectively use JAX's optimizations. Therefore, it is recommended to avoid using the Python for 
loops to effectively maximize JAX's strength. Using ``jax.lax.scan`` function could enable efficiently

> parallel computations.

Reviewer: We should avoid making statements about the strengths/weaknesses of JAX. Here again I think the text can be trimmed down:

"When writing the custom solver, we write and register a version for both NumPy and JAX. The NumPy version uses standard Python loops, whereas the JAX version uses the JAX looping construct `jax.lax.scan`...."

Author: I updated at https://github.com/Qiskit-Extensions/arraylias/pull/25/commits/328ab5b771dca327efab87a6d5d6a871184d06de.

    %timeit unp.array(unp.runge_kutta(init_state, N))

Second case is JAX:

.. jupyter-execute::

    Z = jnp.array([[1,0],[0,-1]])
    X = jnp.array([[0,1],[1,0]])
    init_state = jnp.array([1. + 0j,0. + 0j])
    probabilities = unp.array(unp.runge_kutta(init_state, N))

    T = np.linspace(0,(N-1) * dt,N)

    plt.plot(T, probabilities, label=["0", "1"])
    plt.xlabel('T')
    plt.ylabel('Probability')
    plt.legend()
    plt.show()

We see if we can actually jit.

.. jupyter-execute::

    from functools import partial

    @partial(jax.jit, static_argnums=(1,))
    def solve_with_RungeKutta_jit(init_state, N):
        return unp.array(unp.runge_kutta(init_state, N))

.. jupyter-execute::

    %timeit solve_with_RungeKutta_jit(init_state, N)

By following these steps, we've learned how to leverage Arraylias to 
write versatile numerical code that can efficiently switch between different arrays.
We've also explored how to accelerate our code's execution using JAX's JIT compilation.
By understanding the strengths of different numerical libraries and harnessing their capabilities through Arraylias, we can create high-performance code for various computational tasks. 

> Apply these concepts to your own projects to unlock their full potential.

Reviewer: I don't think this text is necessary - again this is describing general benefits of using arraylias, but this tutorial should just describe the concrete steps necessary to carry out the task at hand.

Author: I updated at https://github.com/Qiskit-Extensions/arraylias/pull/25/commits/328ab5b771dca327efab87a6d5d6a871184d06de.
